zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1009 @@
+            .finish()
+    }
+}
+
+impl From<NullifierDerivingKey> for [u8; 32] {
+    fn from(nk: NullifierDerivingKey) -> [u8; 32] {
+        nk.0.to_bytes()
+    }
+}
+
+impl From<&NullifierDerivingKey> for [u8; 32] {
+    fn from(nk: &NullifierDerivingKey) -> [u8; 32] {
+        nk.0.to_bytes()
+    }
+}
+
+impl From<NullifierDerivingKey> for pallas::Base {
+    fn from(nk: NullifierDerivingKey) -> pallas::Base {
+        nk.0
+    }
+}
+
+impl From<[u8; 32]> for NullifierDerivingKey {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(pallas::Base::from_bytes(&bytes).unwrap())
+    }
+}
+
+impl From<SpendingKey> for NullifierDerivingKey {
+    /// nk = ToBase^Orchard(PRF^expand_sk ([7]))
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#orchardkeycomponents
+    fn from(sk: SpendingKey) -> Self {
+        Self(pallas::Base::from_bytes_wide(&prf_expand(
+            sk.into(),
+            vec![&[7]],
+        )))
+    }
+}
+
+impl Eq for NullifierDerivingKey {}
+
+impl PartialEq<[u8; 32]> for NullifierDerivingKey {
+    fn eq(&self, other: &[u8; 32]) -> bool {
+        <[u8; 32]>::from(*self) == *other
+    }
+}
+
+/// Commit^ivk randomness.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+// XXX: Should this be replaced by commitment::CommitmentRandomness?
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct IvkCommitRandomness(pallas::Scalar);
+
+impl fmt::Debug for IvkCommitRandomness {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("IvkCommitRandomness")
+            .field(&hex::encode(self.0.to_bytes()))
+            .finish()
+    }
+}
+
+impl From<SpendingKey> for IvkCommitRandomness {
+    /// rivk = ToScalar^Orchard(PRF^expand_sk ([8]))
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#orchardkeycomponents
+    fn from(sk: SpendingKey) -> Self {
+        let scalar = pallas::Scalar::from_bytes_wide(&prf_expand(sk.into(), vec![&[8]]));
+
+        Self(scalar)
+    }
+}
+
+impl From<IvkCommitRandomness> for [u8; 32] {
+    fn from(rivk: IvkCommitRandomness) -> Self {
+        rivk.0.into()
+    }
+}
+
+impl From<IvkCommitRandomness> for pallas::Scalar {
+    fn from(rivk: IvkCommitRandomness) -> Self {
+        rivk.0
+    }
+}
+
+impl TryFrom<[u8; 32]> for IvkCommitRandomness {
+    type Error = &'static str;
+
+    fn try_from(bytes: [u8; 32]) -> Result<Self, Self::Error> {
+        let possible_scalar = pallas::Scalar::from_bytes(&bytes);
+
+        if possible_scalar.is_some().into() {
+            Ok(Self(possible_scalar.unwrap()))
+        } else {
+            Err("Invalid pallas::Scalar value")
+        }
+    }
+}
+
+/// Magic human-readable strings used to identify what networks Orchard incoming
+/// viewing keys are associated with when encoded/decoded with bech32.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardinviewingkeyencoding
+mod ivk_hrp {
+    pub const MAINNET: &str = "zivko";
+    pub const TESTNET: &str = "zivktestorchard";
+}
```

Comment:
Similarly, I think the APIs that use these constants could be impacted by Orchard Unified Addresses.

Reply:
They could; this is more up-in-the-air. We might have a "Unified Viewing Key" type.

Reply:
At present these encodings are defined in https://zips.z.cash/protocol/nu5.pdf, if they are impacted by Unified keys we can remove/rework SpendingKey/IncomingViewingKey/FullViewingKey Display/FromStr accordingly.